flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/KnnModelData.java

```diff
@@ -0,0 +1,165 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.classification.knn;
+
+import org.apache.flink.api.common.serialization.Encoder;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.api.java.tuple.Tuple3;
+import org.apache.flink.api.java.typeutils.RowTypeInfo;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.connector.file.src.reader.SimpleStreamFormat;
+import org.apache.flink.core.fs.FSDataInputStream;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.core.memory.DataInputViewStreamWrapper;
+import org.apache.flink.core.memory.DataOutputView;
+import org.apache.flink.core.memory.DataOutputViewStreamWrapper;
+import org.apache.flink.ml.linalg.DenseMatrix;
+import org.apache.flink.ml.linalg.DenseVector;
+import org.apache.flink.ml.linalg.typeinfo.DenseMatrixSerializer;
+import org.apache.flink.ml.linalg.typeinfo.DenseVectorSerializer;
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.api.Schema;
+import org.apache.flink.types.Row;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+
+/** Knn model data, which stores the data used to calculate the distances between nodes. */
+public class KnnModelData {
+    private final List<Tuple3<DenseMatrix, DenseVector, int[]>> dictData;
+    private final Comparator<? super Tuple2<Double, Integer>> comparator;
+
+    /**
+     * Constructor.
+     *
+     * @param list Row list.
+     */
+    public KnnModelData(List<Row> list) {
+        this.dictData = new ArrayList<>(list.size());
+        for (Row row : list) {
+            this.dictData.add(
+                    Tuple3.of(
+                            (DenseMatrix) row.getField(0),
+                            (DenseVector) row.getField(1),
+                            (int[]) row.getField(2)));
+        }
+        comparator = Comparator.comparingDouble(o -> -o.f0);
+    }
+
+    /**
+     * Gets comparator.
+     *
+     * @return Comparator.
+     */
+    public Comparator<? super Tuple2<Double, Integer>> getQueueComparator() {
+        return comparator;
+    }
+
+    /**
+     * Gets dictionary data size.
+     *
+     * @return Dictionary data size.
+     */
+    public Integer getLength() {
+        return dictData.size();
+    }
+
+    public List<Tuple3<DenseMatrix, DenseVector, int[]>> getDictData() {
```

Comment:
Would it be simpler (and also more consistent with other algorithms) to make `dictData` a `public final` field and let caller access this field directly?

Reply:
OK